gather.go

```diff
@@ -90,7 +90,38 @@ func listenUDP(portMax, portMin int, network string, laddr *net.UDPAddr) (*net.U
 	return nil, ErrPort
 }
 
-func gatherCandidatesLocal(a *Agent, networkTypes []NetworkType) {
+// GatherCandidates initiates the trickle based gathering process.
+func (a *Agent) GatherCandidates(urls []*URL, networkTypes []NetworkType) error {
```

Comment:
Should we do our check return `ErrNoOnCandidateHandler` here?

Reply:
The issue here is that this stuff is happening behind this `a.run` task which happens synchronously but by a different goroutine. I originally added ErrNoOnCandidateHandler and was returning it here but because I removed the locks, and put `a.run` I am not sure how to do it anymore.

Reply:
https://github.com/pion/ice/blob/master/candidate_base.go#L109

That is the best we can do for now. I think we want to drop the `a.run` pattern though, someone did a profile and told me it was a major hot spot for them.

Reply:
Fixed!